Runtime error with Vite 8 and vitePreprocess( { script: true}) in Threlte project with @threlte/extras import

### Describe the bug

Hi all,

When I tried to upgrade to Vite 8 to leverage its new build speed increases, I ran into an issue with a production Capacitor/Threlte project where I was getting runtime errors on build due to an erroneous undefined reference to Box3. When I ran `vite build` with minify: false, I could see that one of the references to Box3 wasn't get renamed like the others:

```js
new Box3$1();
new Color$1();
from_html(`<!> <!>`, 1);
from_html(`<!> <!>`, 1);
new Box3();
from_html(`<!> <!>`, 1);
```

I added the `--sourcemap` option to my build and it showed that the errors were coming from `Resize.svelte`, a component that I do not use:

```svelte
<script
  lang="ts"
  module
>
  // reuse box for all instances
  const _box = new Box3() // <-- produces Uncaught (in promise) ReferenceError: Box3 is not defined
</script>

<script lang="ts">
  import {
    isInstanceOf,
    observe,
    T,
    useStage,
    useTask,
    useThrelte,
    type Plugin
  } from '@threlte/core'
  import { Box3, Group } from 'three'
  import InjectPlugin from '../InjectPlugin/InjectPlugin.svelte'
  import type { ResizeProps } from './types.js'

// ...
```

After thorough investigation and troubleshooting, I discovered that the issue resulted from the project's use of Typescript enums. Building with Typescript enums causes a failure in the vite build that says to add `preprocess: vitePreprocess({ script: true })` to `svelte.config.js`.

When this script step was combined with a few basic imports from `@threlte` and `three` , it led to a runtime error that prevented the app from functioning:

```js
 import { OrbitControls, RoundedBoxGeometry } from '@threlte/extras'
 import { T } from "@threlte/core";
 import { Box3, Box3Helper, Mesh } from "three";
```

I was able to reproduce this by installing Threlte on a fresh project, adding some basic imports from @threlte/extras, and building with the `preprocess: vitePreprocess({ script: true })` step.

The workaround was to change all Typescript enums to `as const` and remove the preprocess step from `svelte.config.js`. This removes all runtime errors.

Please let me know if there's any other information I can provide to help resolve this.

### Reproduction URL

https://github.com/fortserious/vitepreprocess-bug

### Reproduction

Run `pnpm install`
Run `pnpm preview`

You will get a 500 error that says Box3 is undefined.

### Logs

```shell
ReferenceError: Box3 is not defined
    <anonymous> Immutable
```

### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
    Memory: 8.51 GB / 31.83 GB
  Binaries:
    Node: 22.14.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.21 - C:\Users\ross\AppData\Roaming\npm\yarn.CMD     
    npm: 10.9.2 - C:\Program Files\nodejs\npm.CMD
    pnpm: 9.12.3 - C:\Users\ross\AppData\Roaming\npm\pnpm.CMD      
    Deno: 1.46.2 - C:\Users\ross\.deno\bin\deno.EXE
  Browsers:
    Chrome: 146.0.7680.165
    Edge: Chromium (140.0.3485.54)
    Firefox: 148.0.2 - C:\Program Files\Mozilla Firefox\firefox.exe
    Internet Explorer: 11.0.19041.5794
  npmPackages:
    @sveltejs/adapter-auto: ^7.0.1 => 7.0.1 
    @sveltejs/adapter-static: ^3.0.10 => 3.0.10 
    @sveltejs/kit: ^2.55.0 => 2.55.0 
    @sveltejs/vite-plugin-svelte: ^7.0.0 => 7.0.0 
    svelte: ^5.55.0 => 5.55.0 
    vite: ^8.0.3 => 8.0.3
```

--- a/.changeset/preprocess-oxc-imports.md
+++ b/.changeset/preprocess-oxc-imports.md
@@ -1,0 +1,5 @@
+---
+'@sveltejs/vite-plugin-svelte': patch
+---
+
+fix: pass `typescript.onlyRemoveTypeImports` to `transformWithOxc` in `vitePreprocess` so that value imports are not dropped when they are only referenced in Svelte template markup

--- a/packages/vite-plugin-svelte/__tests__/preprocess.spec.js
+++ b/packages/vite-plugin-svelte/__tests__/preprocess.spec.js
@@ -108,9 +108,9 @@ describe('vitePreprocess', () => {
 			expect(processed).toBeDefined();
 			// helper is used in script, so it should be preserved by oxc
 			expect(processed.code).toContain('helper');
-			// Component is only used in template (not in script), oxc would strip it
-			// but vitePreprocess should restore it
-			expect(processed.code).toContain("import Component from './Component.svelte'");
+			// Component is only used in template (not in script), but onlyRemoveTypeImports
+			// ensures it is preserved
+			expect(processed.code).toContain('import Component from "./Component.svelte"');
 		});
 
 		it('does not duplicate imports that oxc preserves', async () => {
@@ -125,7 +125,7 @@ describe('vitePreprocess', () => {
 			expect(matches.length).toBe(2); // one in import, one in usage
 		});
 
-		it('does not restore type-only imports', async () => {
+		it('strips type-only imports', async () => {
 			const content = `
   import type { SomeType } from './types.js';
   import Component from './Component.svelte';
@@ -135,27 +135,24 @@ describe('vitePreprocess', () => {
 			expect(processed).toBeDefined();
 			// type import should be correctly stripped
 			expect(processed.code).not.toContain('SomeType');
-			// value import should be restored
-			expect(processed.code).toContain("import Component from './Component.svelte'");
+			// value import should be preserved
+			expect(processed.code).toContain('import Component from "./Component.svelte"');
 		});
 
-		it('restores only stripped named imports when some are used in script', async () => {
+		it('preserves value specifiers and strips type specifiers from mixed imports', async () => {
 			const content = `
-  import { Foo, bar } from './components.js';
-  const x = bar();`;
+  import { Foo, type Bar } from './components.js';
+  const x = 1;`;
 
 			const processed = await script(scriptArgs(content));
 			expect(processed).toBeDefined();
-			// bar is used in script so oxc keeps it
-			expect(processed.code).toContain('bar');
-			// Foo is unused in script but may be used in template - should be restored
+			// Foo is a value import - preserved even though unused in script
 			expect(processed.code).toContain('Foo');
+			// Bar is a type specifier - should be stripped
+			expect(processed.code).not.toContain('Bar');
 		});
 
-		it('restores stripped value identifiers from multi-line mixed imports', async () => {
-			// Simulates the @formatjs/icu-messageformat-parser case:
-			// TYPE is a value used only in template, parse is used in script,
-			// type MessageFormatElement is a type-only specifier
+		it('preserves value identifiers from multi-line mixed imports', async () => {
 			const content = `
   import {
     parse,
@@ -166,32 +163,33 @@ describe('vitePreprocess', () => {
 
 			const processed = await script(scriptArgs(content));
 			expect(processed).toBeDefined();
-			// parse is used in script, should be preserved by oxc
+			// parse is used in script, should be preserved
 			expect(processed.code).toContain('parse');
-			// TYPE is unused in script but used in template - should be restored
+			// TYPE is a value import unused in script - should be preserved
 			expect(processed.code).toContain('TYPE');
-			// type-only specifier should NOT be restored
+			// type-only specifier should be stripped
 			expect(processed.code).not.toContain('MessageFormatElement');
-			// The restored import should reference the correct source
-			expect(processed.code).toContain('@formatjs/icu-messageformat-parser');
 		});
 
-		it('restores default import when identifier only appears as substring elsewhere', async () => {
-			// Regression: "Label" should not be confused with "InputLabels", "labels", etc.
+		it('preserves import even when identifier appears in string literals', async () => {
 			const content = `
-  import Label from './Label.svelte';
-  import type { InputLabels } from './types.js';
-  let labels: InputLabels = {};
-  const defaultLabels = { optionalLabel: 'optional' };`;
+  import Select from '../Select.svelte';
+  import { untrack } from 'svelte';
+  let { withSelect = false } = $props();
+  const defaultLabels = {
+    ariaLabelSelectAllRows: "Select all rows",
+    ariaLabelSelectRow: "Select row",
+  };
+  untrack(() => {});`;
 
 			const processed = await script(scriptArgs(content));
 			expect(processed).toBeDefined();
-			expect(processed.code).toContain("import Label from './Label.svelte'");
+			expect(processed.code).toContain('import Select from "../Select.svelte"');
 		});
 
 		it('skips non-ts script blocks', async () => {
 			const result = await script({
-				content: `import Foo from './Foo.svelte';`,
+				content: "import Foo from './Foo.svelte';",
 				attributes: {},
 				markup: '',
 				filename: `${fixtureDir}/File.svelte`

--- a/packages/vite-plugin-svelte/__tests__/preprocess.spec.js
+++ b/packages/vite-plugin-svelte/__tests__/preprocess.spec.js
@@ -1,4 +1,4 @@
-import { describe, it, expect } from 'vitest';
+import { describe, it, expect, beforeEach } from 'vitest';
 import { vitePreprocess } from '../src/preprocess.js';
 import path from 'node:path';
 import { normalizePath } from 'vite';
@@ -75,6 +75,128 @@ describe('vitePreprocess', () => {
 			expect(dependencies).toBeDefined();
 			expect(dependencies[0]).toBe(path.resolve(fixtureDir, 'foo.scss'));
 			expect(dependencies.length).toBe(1);
+		});
+	});
+
+	describe('script', () => {
+		/** @type {import('svelte/types/compiler/preprocess').Preprocessor} */
+		let script;
+
+		const scriptArgs = (/** @type {string} */ content) => ({
+			content,
+			attributes: { lang: 'ts' },
+			markup: '',
+			filename: `${fixtureDir}/File.svelte`
+		});
+
+		beforeEach(() => {
+			const preprocessorGroup = vitePreprocess({ script: true });
+			script = /** @type {import('svelte/types/compiler/preprocess').Preprocessor} */ (
+				preprocessorGroup.script
+			);
+		});
+
+		it('preserves value imports unused in script but potentially used in template', async () => {
+			const content = `
+  import Component from './Component.svelte';
+  import { helper } from './utils.js';
+  interface Props { value: string; }
+  let { value } = $props();
+  const x = helper(value);`;
+
+			const processed = await script(scriptArgs(content));
+			expect(processed).toBeDefined();
+			// helper is used in script, so it should be preserved by oxc
+			expect(processed.code).toContain('helper');
+			// Component is only used in template (not in script), oxc would strip it
+			// but vitePreprocess should restore it
+			expect(processed.code).toContain("import Component from './Component.svelte'");
+		});
+
+		it('does not duplicate imports that oxc preserves', async () => {
+			const content = `
+  import { writable } from 'svelte/store';
+  const store = writable(0);`;
+
+			const processed = await script(scriptArgs(content));
+			expect(processed).toBeDefined();
+			// writable is used in script, should appear exactly once as an import
+			const matches = processed.code.match(/writable/g);
+			expect(matches.length).toBe(2); // one in import, one in usage
+		});
+
+		it('does not restore type-only imports', async () => {
+			const content = `
+  import type { SomeType } from './types.js';
+  import Component from './Component.svelte';
+  let x: SomeType;`;
+
+			const processed = await script(scriptArgs(content));
+			expect(processed).toBeDefined();
+			// type import should be correctly stripped
+			expect(processed.code).not.toContain('SomeType');
+			// value import should be restored
+			expect(processed.code).toContain("import Component from './Component.svelte'");
+		});
+
+		it('restores only stripped named imports when some are used in script', async () => {
+			const content = `
+  import { Foo, bar } from './components.js';
+  const x = bar();`;
+
+			const processed = await script(scriptArgs(content));
+			expect(processed).toBeDefined();
+			// bar is used in script so oxc keeps it
+			expect(processed.code).toContain('bar');
+			// Foo is unused in script but may be used in template - should be restored
+			expect(processed.code).toContain('Foo');
+		});
+
+		it('restores stripped value identifiers from multi-line mixed imports', async () => {
+			// Simulates the @formatjs/icu-messageformat-parser case:
+			// TYPE is a value used only in template, parse is used in script,
+			// type MessageFormatElement is a type-only specifier
+			const content = `
+  import {
+    parse,
+    TYPE,
+    type MessageFormatElement,
+  } from "@formatjs/icu-messageformat-parser";
+  const ast = parse("hello");`;
+
+			const processed = await script(scriptArgs(content));
+			expect(processed).toBeDefined();
+			// parse is used in script, should be preserved by oxc
+			expect(processed.code).toContain('parse');
+			// TYPE is unused in script but used in template - should be restored
+			expect(processed.code).toContain('TYPE');
+			// type-only specifier should NOT be restored
+			expect(processed.code).not.toContain('MessageFormatElement');
+			// The restored import should reference the correct source
+			expect(processed.code).toContain('@formatjs/icu-messageformat-parser');
+		});
+
+		it('restores default import when identifier only appears as substring elsewhere', async () => {
+			// Regression: "Label" should not be confused with "InputLabels", "labels", etc.
+			const content = `
+  import Label from './Label.svelte';
+  import type { InputLabels } from './types.js';
+  let labels: InputLabels = {};
+  const defaultLabels = { optionalLabel: 'optional' };`;
+
+			const processed = await script(scriptArgs(content));
+			expect(processed).toBeDefined();
+			expect(processed.code).toContain("import Label from './Label.svelte'");
+		});
+
+		it('skips non-ts script blocks', async () => {
+			const result = await script({
+				content: `import Foo from './Foo.svelte';`,
+				attributes: {},
+				markup: '',
+				filename: `${fixtureDir}/File.svelte`
+			});
+			expect(result).toBeUndefined();
 		});
 	});
 });

--- a/packages/vite-plugin-svelte/src/preprocess.js
+++ b/packages/vite-plugin-svelte/src/preprocess.js
@@ -50,24 +50,130 @@ function viteScript() {
 			const { code, map } = await transformWithOxc(content, filename, {
 				lang,
 				target: 'esnext'
-				// TODO, how to pass tsconfig compilerOptions (or not needed as config is loaded for file
-				/*tsconfigRaw: {
-					compilerOptions: {
-						// svelte typescript needs this flag to work with type imports
-						importsNotUsedAsValues: 'preserve',
-						preserveValueImports: true
-					}
-				}*/
 			});
+
+			// oxc strips imports it considers unused, but some may be referenced in the Svelte
+			// template which oxc cannot see. Detect stripped value imports and re-add them.
+			const restoredCode = restoreStrippedImports(content, code);
 
 			mapToRelative(map, filename);
 
 			return {
-				code,
+				code: restoredCode,
 				map
 			};
 		}
 	};
+}
+
+/**
+ * @typedef {{ kind: 'default' | 'named' | 'namespace', local: string, imported?: string }} ImportedId
+ */
+
+/**
+ * Extracts imported value identifiers from an import statement.
+ * Skips type-only specifiers (e.g., `type Foo` inside `import { type Foo, Bar }`).
+ * @param {string} importStatement
+ * @returns {{ ids: ImportedId[], source: string } | null}
+ */
+function getImportedIdentifiers(importStatement) {
+	// skip `import type { ... }` or `import type Foo` entirely
+	if (/\bimport\s+type\b/.test(importStatement)) return null;
+
+	const sourceMatch = importStatement.match(/from\s+['"]([^'"]+)['"]/);
+	if (!sourceMatch) return null;
+	const source = sourceMatch[1];
+
+	/** @type {ImportedId[]} */
+	const ids = [];
+
+	// default import: import Foo from '...'
+	const defaultMatch = importStatement.match(/import\s+([A-Za-z_$][\w$]*)\s+from/);
+	if (defaultMatch) ids.push({ kind: 'default', local: defaultMatch[1] });
+
+	// named imports: import { a, b as c, type D } from '...'
+	const namedMatch = importStatement.match(/import\s*(?:[A-Za-z_$][\w$]*\s*,\s*)?\{([^}]+)\}/);
+	if (namedMatch) {
+		for (const specifier of namedMatch[1].split(',')) {
+			const trimmed = specifier.trim();
+			if (!trimmed) continue;
+			// skip type-only specifiers: `type Foo` or `type Foo as Bar`
+			if (/^type\s+/.test(trimmed)) continue;
+			const asMatch = trimmed.match(/^(\S+)\s+as\s+(\S+)$/);
+			if (asMatch) {
+				ids.push({ kind: 'named', imported: asMatch[1], local: asMatch[2] });
+			} else {
+				ids.push({ kind: 'named', local: trimmed });
+			}
+		}
+	}
+
+	// namespace import: import * as X from '...'
+	const nsMatch = importStatement.match(/import\s*\*\s*as\s+(\S+)/);
+	if (nsMatch) ids.push({ kind: 'namespace', local: nsMatch[1] });
+
+	return ids.length > 0 ? { ids, source } : null;
+}
+
+const importFromRe = /^\s*import\s+[\s\S]+?\s+from\s+['"][^'"]+['"]\s*;?\s*$/gm;
+
+/**
+ * Compares original script content with oxc output and re-adds value imports
+ * that were stripped because they appeared unused within the script block.
+ * These imports may be referenced in the Svelte template.
+ * Handles partial stripping: if an import has both used and unused identifiers,
+ * only the stripped ones are re-added.
+ * @param {string} original - original script block content
+ * @param {string} transformed - oxc-transformed output
+ * @returns {string}
+ */
+function restoreStrippedImports(original, transformed) {
+	const originalImports = original.match(importFromRe);
+	if (!originalImports) return transformed;
+
+	/** @type {string[]} */
+	const restoreStatements = [];
+
+	for (const imp of originalImports) {
+		const parsed = getImportedIdentifiers(imp);
+		if (!parsed) continue;
+
+		const missingIds = parsed.ids.filter(
+			(id) => !new RegExp(`\\b${id.local}\\b`).test(transformed)
+		);
+		if (missingIds.length === 0) continue;
+
+		// If ALL identifiers are missing, restore the original import as-is (single-lined)
+		if (missingIds.length === parsed.ids.length) {
+			restoreStatements.push(imp.replace(/\s*\n\s*/g, ' ').trim());
+			continue;
+		}
+
+		// Partial strip: rebuild import for only the missing identifiers
+		const defaultIds = missingIds.filter((id) => id.kind === 'default');
+		const namedIds = missingIds.filter((id) => id.kind === 'named');
+		const nsIds = missingIds.filter((id) => id.kind === 'namespace');
+
+		/** @type {string[]} */
+		const parts = [];
+		if (defaultIds.length > 0) parts.push(defaultIds[0].local);
+		if (namedIds.length > 0) {
+			const specifiers = namedIds.map((id) =>
+				id.imported ? `${id.imported} as ${id.local}` : id.local
+			);
+			parts.push(`{ ${specifiers.join(', ')} }`);
+		}
+		if (nsIds.length > 0) parts.push(`* as ${nsIds[0].local}`);
+
+		if (parts.length > 0) {
+			restoreStatements.push(`import ${parts.join(', ')} from '${parsed.source}';`);
+		}
+	}
+
+	if (restoreStatements.length > 0) {
+		return restoreStatements.join('\n') + '\n' + transformed;
+	}
+	return transformed;
 }
 
 /**

--- a/packages/vite-plugin-svelte/src/preprocess.js
+++ b/packages/vite-plugin-svelte/src/preprocess.js
@@ -49,131 +49,22 @@ function viteScript() {
 			const lang = /** @type {'ts'} */ (attributes.lang);
 			const { code, map } = await transformWithOxc(content, filename, {
 				lang,
-				target: 'esnext'
+				target: 'esnext',
+				// oxc strips value imports it considers unused, but in Svelte files
+				// these imports may be referenced in the template which oxc cannot see.
+				// onlyRemoveTypeImports tells oxc to only strip type-only imports
+				// (import type {...}, import { type X }) and preserve all value imports.
+				typescript: { onlyRemoveTypeImports: true }
 			});
-
-			// oxc strips imports it considers unused, but some may be referenced in the Svelte
-			// template which oxc cannot see. Detect stripped value imports and re-add them.
-			const restoredCode = restoreStrippedImports(content, code);
 
 			mapToRelative(map, filename);
 
 			return {
-				code: restoredCode,
+				code,
 				map
 			};
 		}
 	};
-}
-
-/**
- * @typedef {{ kind: 'default' | 'named' | 'namespace', local: string, imported?: string }} ImportedId
- */
-
-/**
- * Extracts imported value identifiers from an import statement.
- * Skips type-only specifiers (e.g., `type Foo` inside `import { type Foo, Bar }`).
- * @param {string} importStatement
- * @returns {{ ids: ImportedId[], source: string } | null}
- */
-function getImportedIdentifiers(importStatement) {
-	// skip `import type { ... }` or `import type Foo` entirely
-	if (/\bimport\s+type\b/.test(importStatement)) return null;
-
-	const sourceMatch = importStatement.match(/from\s+['"]([^'"]+)['"]/);
-	if (!sourceMatch) return null;
-	const source = sourceMatch[1];
-
-	/** @type {ImportedId[]} */
-	const ids = [];
-
-	// default import: import Foo from '...'
-	const defaultMatch = importStatement.match(/import\s+([A-Za-z_$][\w$]*)\s+from/);
-	if (defaultMatch) ids.push({ kind: 'default', local: defaultMatch[1] });
-
-	// named imports: import { a, b as c, type D } from '...'
-	const namedMatch = importStatement.match(/import\s*(?:[A-Za-z_$][\w$]*\s*,\s*)?\{([^}]+)\}/);
-	if (namedMatch) {
-		for (const specifier of namedMatch[1].split(',')) {
-			const trimmed = specifier.trim();
-			if (!trimmed) continue;
-			// skip type-only specifiers: `type Foo` or `type Foo as Bar`
-			if (/^type\s+/.test(trimmed)) continue;
-			const asMatch = trimmed.match(/^(\S+)\s+as\s+(\S+)$/);
-			if (asMatch) {
-				ids.push({ kind: 'named', imported: asMatch[1], local: asMatch[2] });
-			} else {
-				ids.push({ kind: 'named', local: trimmed });
-			}
-		}
-	}
-
-	// namespace import: import * as X from '...'
-	const nsMatch = importStatement.match(/import\s*\*\s*as\s+(\S+)/);
-	if (nsMatch) ids.push({ kind: 'namespace', local: nsMatch[1] });
-
-	return ids.length > 0 ? { ids, source } : null;
-}
-
-const importFromRe = /^\s*import\s+[\s\S]+?\s+from\s+['"][^'"]+['"]\s*;?\s*$/gm;
-
-/**
- * Compares original script content with oxc output and re-adds value imports
- * that were stripped because they appeared unused within the script block.
- * These imports may be referenced in the Svelte template.
- * Handles partial stripping: if an import has both used and unused identifiers,
- * only the stripped ones are re-added.
- * @param {string} original - original script block content
- * @param {string} transformed - oxc-transformed output
- * @returns {string}
- */
-function restoreStrippedImports(original, transformed) {
-	const originalImports = original.match(importFromRe);
-	if (!originalImports) return transformed;
-
-	/** @type {string[]} */
-	const restoreStatements = [];
-
-	for (const imp of originalImports) {
-		const parsed = getImportedIdentifiers(imp);
-		if (!parsed) continue;
-
-		const missingIds = parsed.ids.filter(
-			(id) => !new RegExp(`\\b${id.local}\\b`).test(transformed)
-		);
-		if (missingIds.length === 0) continue;
-
-		// If ALL identifiers are missing, restore the original import as-is (single-lined)
-		if (missingIds.length === parsed.ids.length) {
-			restoreStatements.push(imp.replace(/\s*\n\s*/g, ' ').trim());
-			continue;
-		}
-
-		// Partial strip: rebuild import for only the missing identifiers
-		const defaultIds = missingIds.filter((id) => id.kind === 'default');
-		const namedIds = missingIds.filter((id) => id.kind === 'named');
-		const nsIds = missingIds.filter((id) => id.kind === 'namespace');
-
-		/** @type {string[]} */
-		const parts = [];
-		if (defaultIds.length > 0) parts.push(defaultIds[0].local);
-		if (namedIds.length > 0) {
-			const specifiers = namedIds.map((id) =>
-				id.imported ? `${id.imported} as ${id.local}` : id.local
-			);
-			parts.push(`{ ${specifiers.join(', ')} }`);
-		}
-		if (nsIds.length > 0) parts.push(`* as ${nsIds[0].local}`);
-
-		if (parts.length > 0) {
-			restoreStatements.push(`import ${parts.join(', ')} from '${parsed.source}';`);
-		}
-	}
-
-	if (restoreStatements.length > 0) {
-		return restoreStatements.join('\n') + '\n' + transformed;
-	}
-	return transformed;
 }
 
 /**

